dcos_launch/terraform.py

```diff
@@ -0,0 +1,283 @@
+                            'to it i.e. "ssh-add /path/to/key.pem". ssh-agent usage is specific to terraform, not '
+                            'dcos-launch.')
+
+            module = 'github.com/dcos/{}?ref={}/{}'.format(
+                'terraform-dcos-enterprise' if self.config['dcos-enterprise'] else 'terraform-dcos',
+                self.config['terraform_dcos_enterprise_version'] if self.config['dcos-enterprise'] else
+                self.config['terraform_dcos_version'], self.config['platform'])
+
+            # Converting our YAML config to the required format. You can find an example of that format in the
+            # Advance YAML Configuration" section here:
+            # https://github.com/mesosphere/terraform-dcos-enterprise/tree/master/aws
+            with open(self.cluster_profile_path, 'w') as file:
+                for k, v in self.config['terraform_config'].items():
+                    file.write(k + ' = ')
+                    if type(k) is dict:
+                        file.write('<<EOF\n{}\nEOF\n'.format(yaml.dump(v)))
+                    else:
+                        file.write('"{}"\n'.format(v))
+            subprocess.run([self.terraform_cmd(), 'init', '-from-module', module], cwd=self.init_dir,
+                           check=True, stderr=subprocess.STDOUT)
+            subprocess.run([self.terraform_cmd(), 'apply', '-auto-approve', '-var-file', self.cluster_profile_path],
+                           cwd=self.init_dir, check=True, stderr=subprocess.STDOUT, env=os.environ)
+        except Exception as e:
+            self.config['create_exception'] = e
+        return self.config
+
+    def _install_terraform(self):
+        download_path = os.path.join(self.dcos_launch_root_dir, 'terraform.zip')
+        try:
+            with open(download_path, 'wb') as f:
+                log.info('Downloading...')
+                r = requests.get(self.config['terraform_tarball_url'])
+                for chunk in r.iter_content(1024):
+                    f.write(chunk)
+            with zipfile.ZipFile(download_path, 'r') as tfm_zip:
+                tfm_zip.extractall(self.dcos_launch_root_dir)
+            # setting terraform binary permissions to: execute only, by file owner only
+            os.chmod(self.terraform_binary, 0o100)
+        finally:
+            os.remove(download_path)
+        log.info('Terraform installation complete.')
+
+    def wait(self):
+        """ Nothing to do here because unlike the other launchers, create() runs and also waits for a subprocess to
+        finish instead of just sending an http request to a provider.
+        """
+        pass
+
+    def delete(self):
+        # delete the cluster
+        subprocess.run([self.terraform_cmd(), 'destroy', '-force', '-var-file', self.cluster_profile_path],
+                       cwd=self.init_dir, check=True, stderr=subprocess.STDOUT, env=os.environ)
+        # remove the init dir
+        shutil.rmtree(self.init_dir, ignore_errors=True)
```

Comment:
Could this potentially wipe out the current, working directory? (I may not understand `init_dir`'s value)

Reply:
init-dir: terraform directory/environment created only for this cluster